unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt

```diff
@@ -53,25 +53,37 @@
 #
 # For each character in <unicodeSet>, verify that the result of applying the left <props>
 # is (=|≠) the result of applying the right <props>.
-#   <props> is of the form (<unicodeSet> | <prop>) ("*" (<unicodeSet> | <prop>))?
+#   <props> is of the form (<unicodeSet> | <prop>) ("*" (<unicodeSet> | <prop> | <sequenceTransformation>))?
 #   It is the functional composition of the properties applied to strings, whereby
 #   <unicodeSet> is used to filter the result.
 #   <prop> for a string property is applied to each character, and the result concatenated
 #       That is, cf("A1") is cf("A")+cf("1") = "a1"
-#   <prop> for an enumerated property, is applied to each character, and the result is a concatenated set.
-#       That is, gc("A1") is gc("A")+gc("1") = "Uppercase_LetterDecimal_Number"
+#   <prop> for an enumerated property is applied to each character, and the result is sequence of
+#   strings.
+#       That is, gc("A1") is [gc("A"), gc("1")] = ["Uppercase_Letter", "Decimal_Number"]
+#   <sequenceTransformation> may be applied to a seqence or strings or to a string.  On a string it
+#   operates on the code points and returns a string.
+#       The available operations are:
+#           (append <string>)
+#           (prepend <string>)
+#           (take <n>)
+#           (drop <n>)
+#           (delete-adjacent-duplicates)
+#           (string-join)
+#   The result of the <props> must be a string, a single-string sequence, or an empty sequence.
 #
 #   Example: for <props> of bc * \P{bc=NSM} * cf * dm, the result applied to Å (angstrom sign) are:
 #       bc * \P{bc=NSM} * cf * dm ("Å")
 #       bc * \P{bc=NSM} * cf ("A" + umlaut)
 #       bc * \P{bc=NSM} ("a" + umlaut)
 #       bc ("a")
-#       "Left"
+#       "Left_To_Right"
 #
-#   Example: In \p{dt=canonical} bc * \P{bc=NSM} * dm = bc * \P{bc=NSM}
+#   Example: In \p{dt=canonical} (delete-adjacent-duplicates) * bc * \P{bc=NSM} * dm = bc * \P{bc=NSM}
 #       This examines only those characters that have canonical compositions. For each such character X
-#       it gets the decomposition mapping of X, then filters out all NSM characters, then gets the Bidi_Class.
-#       It then tests that against the result of filtering out NSM characters from X, then getting the BIDI_Class.
+#       it gets the decomposition mapping of X, then filters out all NSM characters, then gets the Bidi_Class,
+#       then deduplicates runs of the same Bidi_Class.
+#       It then compares that with the result of filtering out NSM characters from X, then getting the Bidi_Class.
```

Comment:
What happens if filtering out NSMs from x yields an empty string? What does applying bc to that do?

Reply:
It returns the sequence of bc for every code point in the empty string; that is the empty sequence.